Added check for existence of Location HTTP header before using it

ffeldhaus opened the pull request:
## PR Summary

Fix #6543 

The [HTTP RFC does not require a Location header to be present for redirects](https://tools.ietf.org/html/rfc7231#section-6.4), thus it is required to check if the Location header is returned before using it. This pull request adds a check for the Location header.

A test has been implemented to ensure that the issue does not show up again.

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Change is not breaking](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` to the beginning of the title and remove the prefix when the PR is ready.
- **User-facing changes**
    - [X] Not Applicable
- **Testing - New and feature**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
        - [x] [Add `[feature]` if the change is significant or affects feature tests](https://github.com/PowerShell/PowerShell/blob/master/docs/testing-guidelines/testing-guidelines.md#requesting-additional-tests-for-a-pr)

ffeldhaus committed:
Check existence of Location header before using it

markekraus commented:
Hi @ffeldhaus ! Can you please add a test for this for bot `Invoke-WebRequest` and `Invoke-RestMethod` so we can ensure future changes don't cause a regression?

Also, for this PR, make sure you include `[feature]` in the commit messages to ensure the web cmdlet tests run.

Thanks!

markekraus commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs:
```diff
@@ -1287,7 +1287,7 @@ internal virtual HttpResponseMessage GetResponse(HttpClient client, HttpRequestM
             HttpResponseMessage response = client.SendAsync(request, HttpCompletionOption.ResponseHeadersRead, _cancelToken.Token).GetAwaiter().GetResult();
 
-            if (stripAuthorization && IsRedirectCode(response.StatusCode))
+            if (stripAuthorization && IsRedirectCode(response.StatusCode) && response.Headers.Location)
```
should be `response.Headers.Location != null`

ffeldhaus committed:
Fixed Location header check to validate if it is not null

ffeldhaus commented:
Hi @markekraus thanks for the suggestion. I already fixed the initial commit, I should have tested it before submitting it. I now confirmed that the fix is working and committed the correct fix.

I would like to add a test, but as far as I understand WebListener, it always returns the Location header for Redirects. What would be the best way to make WebListener not return the Location header? If you can help me with the WebListener, I'd be glad to commit a test for both Cmdlets.

markekraus commented:
@ffeldhaus Have a look at the `/Response/` endpoint https://github.com/PowerShell/PowerShell/tree/master/test/tools/WebListener#response 
You can have it return a redirect code without a Location header.

something like this

```powershell
$uri = Get-WebListenerUrl -Test Response -Query @{statuscode = 301}
{ Invoke-WebRequest -uri $uri -Headers @{Authorization = 'foo'} -ErroractionStop } |
    Should -Throw -ErrorId 'WebCmdletWebResponseException,Microsoft.PowerShell.Commands.InvokeWebRequestCommand'
```

ffeldhaus committed:
[Feature] Added tests to validate that WebCmdlets correctly handle redirects without location header

markekraus commented:
@ffeldhaus It looks like an unintentional change was made in your last commit that deleted an unrelated file. Can you please revert the file?

ffeldhaus committed:
Revert unrelated file about_PSReadline.help.txt

ffeldhaus committed:
[feature] Changed status code from string to value to allow WebListener to respond correctly

ffeldhaus committed:
[feature] Changed status code from string to int and using -Match instead of -Be to be more flexible

The WebListener requires the status code to be an int, therefore the test was changed to only use the integer form of the Status Code. As the Exception spells out the error message with spaces in the Status Code string, the -Be was changed to -Match and the String form of the Status Code removed.

ffeldhaus commented:
I changed the status code to int and are using -Match now isntead of -Be for checking the Exception message. The reason is, that the Exception Message contains the Status Code with a space, e.g. `Temporary Redirect` and not `TemporaryRedirect` which is provided in `$redirectTests`.

The tests are still failing because `$redirectTests` contains an entry `relative` which is not a valid Status Code. I'm not sure how to handle this. In my opinion it would be better to not include `relative` in the list of `$redirectTests` or at least separate it into Status Codes to test for redirect and an additional `relative` redirect test.

markekraus commented:
@ffeldhaus for the `relative` issue, you could do 

```powershell
# Skip relative test as it is not a valid response type.
if ($redirectType -eq 'relative'){ return }
```

For the error message. I'm not certain we should even test it. Ideally, that message would be localized so it would fail in another language. I think the other tests are sufficient to show we are getting the error condition we desire and not the unrelated one.

ffeldhaus committed:
[feature] Skipping relative test as it is not a valid response type and removed testing for Exception message as the other tests are sufficient

ffeldhaus commented:
@markekraus Thanks for all the help! I'm a bit annoyed at myself for doing quite a few mistakes in such an easy fix and that you basically had to tell me what to do. For the next time, I hope I'm now better equipped to make a change.

I think the pull request should be good to merge now.

markekraus commented on test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1:
```diff
@@ -822,6 +822,24 @@ Describe "Invoke-WebRequest tests" -Tags "Feature" {
+
+        # PowerShell should throw an HTTP Response Exception
+        # for a redirect response which does not contain a Location header
+        # the correct redirect status code should be included in the exception message
```
We should include a note about  this requiring a Authorization header. Something like

```powershell
# When an Authorization request header is present,
# and -PreserveAuthorizationOnRedirect is not present,
# PowerShell should throw an HTTP Response Exception
# for a redirect response which does not contain a Location response header.
# The correct redirect status code should be included in the exception.
```

It's a mouth full, but at least people will understand why we are doing what seems like a superfluous set of tests.

markekraus commented on test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1:
```diff
@@ -2137,6 +2155,24 @@ Describe "Invoke-RestMethod tests" -Tags "Feature" {
+
+        # PowerShell should throw an HTTP Response Exception
+        # for a redirect response which does not contain a Location header
+        # the correct redirect status code should be included in the exception message
```
Same as above.

markekraus reviewed (changes requested):
One more change and I think we're good.

ffeldhaus committed:
[feature] Clarified test description and added more information on what is tested

markekraus reviewed (approved):
LGTM

iSazonov commented on test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1:
```diff
@@ -822,6 +822,27 @@ Describe "Invoke-WebRequest tests" -Tags "Feature" {
         $response.Content.Method | Should -Be $redirectedMethod
     }
 
+    It "Validates Invoke-WebRequest handles responses without Location header correctly for requests with Authorization header and redirect: <redirectType>" -TestCases $redirectTests {
```
It is good for comment but very long for header. Please make shorter header and add the comment.
Below too.

markekraus replied:
@TravisEz13 I get conflicting feedback on that. For now, I think it's best if this follows the code around it. When I refactor the tests, I will address this globally.

ffeldhaus replied:
I added the whitespace and shortened the test description by one word. It's now shorter than the surrounding tests. I agree that `Validates` is redundant, but as all the other tests include it, I include it for consistency. 

I don't agree that a test header should be short. It really must describe the test in a good way and removing anything (apart from `Validates`) would change the meaning of the test.

@markekraus already stated that he wants to cleanup the tests and he can consistently remove `Validates` for all tests.

iSazonov commented on test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1:
```diff
@@ -822,6 +822,27 @@ Describe "Invoke-WebRequest tests" -Tags "Feature" {
+    It "Validates Invoke-WebRequest handles responses without Location header correctly for requests with Authorization header and redirect: <redirectType>" -TestCases $redirectTests {
+        param($redirectType, $redirectedMethod)
+        # Skip relative test as it is not a valid response type.
+        if ($redirectType -eq 'relative'){ return }
```
Please add space after `)` - `if ($redirectType -eq 'relative') { return }`

iSazonov commented on test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1:
```diff
@@ -2137,6 +2158,27 @@ Describe "Invoke-RestMethod tests" -Tags "Feature" {
+    It "Validates Invoke-RestMethod handles responses without Location header correctly for requests with Authorization header and redirect: <redirectType>" -TestCases $redirectTests {
+        param($redirectType, $redirectedMethod)
+        # Skip relative test as it is not a valid response type.
+        if ($redirectType -eq 'relative'){ return }
```
Please add space after `)`.

ffeldhaus committed:
[feature] Shortened test description by one word and added whitespace

iSazonov commented:
@ffeldhaus Thanks for your contribution! Welcome back with new ones!
